Sources/AsyncHTTPClient/Utils.swift

```diff
@@ -52,174 +53,207 @@ public final class HTTPClientCopyingDelegate: HTTPClientResponseDelegate {
-            return NIOClientTCPBootstrap(self, tls: tlsProvider)
+            sslContext = eventLoop.makeSucceededFuture(nil)
         }
-    }
-}
 
-extension NIOClientTCPBootstrap {
-    /// create a TCP Bootstrap based off what type of `EventLoop` has been passed to the function.
-    fileprivate static func makeBootstrap(
-        on eventLoop: EventLoop,
-        host: String,
-        requiresTLS: Bool,
-        configuration: HTTPClient.Configuration
-    ) throws -> NIOClientTCPBootstrap {
-        var bootstrap: NIOClientTCPBootstrap
-        #if canImport(Network)
-            // if eventLoop is compatible with NIOTransportServices create a NIOTSConnectionBootstrap
-            if #available(OSX 10.14, iOS 12.0, tvOS 12.0, watchOS 6.0, *), let tsBootstrap = NIOTSConnectionBootstrap(validatingGroup: eventLoop) {
-                // if there is a proxy don't create TLS provider as it will be added at a later point
-                if configuration.proxy != nil {
-                    bootstrap = NIOClientTCPBootstrap(tsBootstrap, tls: NIOInsecureNoTLS())
+        let bootstrap = NIOClientTCPBootstrap.makeAndConfigureBootstrap(on: channelEventLoop,
+                                                                        host: key.host,
+                                                                        port: key.port,
+                                                                        requiresTLS: requiresTLS,
+                                                                        configuration: configuration,
+                                                                        sslContextCache: sslContextCache,
+                                                                        logger: logger)
+        return bootstrap.flatMap { bootstrap -> EventLoopFuture<Channel> in
+            let channel: EventLoopFuture<Channel>
+            switch key.scheme {
+            case .http, .https:
+                let address = HTTPClient.resolveAddress(host: key.host, port: key.port, proxy: configuration.proxy)
+                channel = bootstrap.connect(host: address.host, port: address.port)
+            case .unix, .http_unix, .https_unix:
+                channel = bootstrap.connect(unixDomainSocketPath: key.unixPath)
+            }
+
+            return channel.flatMap { channel -> EventLoopFuture<(Channel, NIOSSLContext?)> in
+                sslContext.map { sslContext -> (Channel, NIOSSLContext?) in
+                    (channel, sslContext)
+                }
+            }.flatMap { channel, sslContext in
+                configureChannelPipeline(channel,
+                                         isNIOTS: bootstrap.isNIOTS,
+                                         sslContext: sslContext,
+                                         configuration: configuration,
+                                         key: key)
```

Comment:
Nevermind. I thought we can collapse the `sslContext.map` into the following `flatMap`, but that doesn't work, since the `sslContext` is a `EventLoopFuture<NIOSSLContext?>` first and actual `NIOSSLContext?` later.